### PR TITLE
Move to_sparse_csr to C++

### PR DESCRIPTION
**Pull request by cpuhrsch**
Allows use of to_sparse_csr from C++

### CODE REVIEW

**Comment by cpuhrsch** [comment]
Currently getting error
```
ERROR [0.006s]: test_autograd_dense_output_mv_cpu_float64 (__main__.TestSparseCSRCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 389, in instantiated_test
    raise rte
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 761, in test_wrapper
    return test(*args, **kwargs)
  File "test_sparse_csr.py", line 1543, in test_autograd_dense_output
    self.assertTrue(torch.autograd.gradcheck(fn, sample.args, fast_mode=True))
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 1409, in gradcheck
    return _gradcheck_helper(**args)
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 1426, in _gradcheck_helper
    check_undefined_grad=check_undefined_grad)
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 1071, in _gradcheck_real_imag
    rtol, atol, check_grad_dtypes, nondet_tol)
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 1300, in _fast_gradcheck
    analytical_vJu = _get_analytical_vJu_backward_mode(inputs, outputs, nondet_tol, check_grad_dtypes, all_v, all_u_dense)
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 556, in _get_analytical_vJu_backward_mode
    fast_mode=True, v=v)
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 528, in _check_analytical_jacobian_attributes
    vjps1 = _get_analytical_vjps_wrt_specific_output(vjp_fn, output.clone(), v)
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 633, in _get_analytical_vjps_wrt_specific_output
    grad_inputs = vjp_fn(v.reshape(sample_output.shape))
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 525, in vjp_fn
    retain_graph=True, allow_unused=True)
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/__init__.py", line 277, in grad
    allow_unused, accumulate_grad=False)  # Calls into the C++ engine to run the backward pass
RuntimeError: Function MvBackward0 returned an invalid gradient at index 0 - expected layout SparseCsr but got Strided
```

**Comment by IvanYashchuk** [comment]
Okay, tests started to fail because previously the behavior was:
```py
In [1]: import torch

In [2]: a = torch.randn(3, 3, requires_grad=True)

In [3]: aa = a.to_sparse_csr()

In [4]: aa
Out[4]:
tensor(crow_indices=tensor([0, 3, 6, 9]),
       col_indices=tensor([0, 1, 2, 0, 1, 2, 0, 1, 2]),
       values=tensor([-1.1579,  0.7685, -1.1372,  1.0388, -1.0908, -0.0759,
                       1.1043, -0.2585,  1.1056]), size=(3, 3), nnz=9,
       layout=torch.sparse_csr)

In [5]: aa.requires_grad
Out[5]: False
```
And with this PR the "requires_grad" property propagates through the `.to_sparse_csr()` call:
```py
In [1]: import torch

In [2]: a = torch.randn(3, 3, requires_grad=True)

In [3]: a
Out[3]:
tensor([[-2.6340, -0.4176,  0.5212],
        [-1.7476,  1.3983,  0.2506],
        [ 0.7960, -0.7103,  1.3390]], requires_grad=True)

In [4]: aa = a.to_sparse_csr()

In [5]: aa
Out[5]:
tensor(crow_indices=tensor([0, 3, 6, 9]),
       col_indices=tensor([0, 1, 2, 0, 1, 2, 0, 1, 2]),
       values=tensor([-2.6340, -0.4176,  0.5212, -1.7476,  1.3983,  0.2506,
                       0.7960, -0.7103,  1.3390]), size=(3, 3), nnz=9,
       layout=torch.sparse_csr, grad_fn=<NotImplemented>)

In [6]: aa.grad_fn
Out[6]: <NotImplemented at 0x7fd3504c1250>

In [7]: aa.requires_grad
Out[7]: True
```

The fix would be to set the requires_grad property to False on the CSR input in autograd tests.

**Comment by cpuhrsch** [comment]
@IvanYashchuk - seems like it's passing now. Can you give it another look?

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -1574,6 +1574,9 @@
 - name: to_sparse.sparse_dim(Tensor self, int sparse_dim) -> Tensor
   self: grad.to_dense()
 
+- name: to_sparse_csr(Tensor self) -> Tensor
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:1577]: Why do you have this formula if you don't test it?

Comment (cpuhrsch) [reply tools/autograd/derivatives.yaml:1577]: I thought it would fix the issue before Ivan diagnosed it. I'm adding an explicit autograd test now.

Comment (IvanYashchuk) [reply tools/autograd/derivatives.yaml:1577]: CSR layout is not yet supported by autograd. So the test won't pass. It's better to remove this formula for now, we'll add it later once autograd works.

**Review by ngimel — APPROVED** [review]

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**